### PR TITLE
Fused SelfAttention and Cross Attention Decoder

### PR DESCRIPTION
**Pull request by pbontrager**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

This adds a prototype implementation of a flamingo model. This introduces fused layers and fused embeddings for combining cross attention layers with a decoder language model backbone. In the included component builders, this PR uses Llama3 as the reference.

#### Changelog
- Flamingo component builders - llama3 based decoder component
- modules/model_fusion - modules for fusing layers between pre-trained model, loosely follows peft design

Out of scope: flamingo encoder, utilities for manipulating fused model parameters

#### Test plan
This is a new experimental component builder, it will be tested once a model builder is put together.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

File: torchtune/modules/transformer.py [hunk 1]
```diff
@@ -57,27 +65,144 @@ def forward(
...
+                During inference, this indicates the position of the current token.
+                If none, assume the index of the token is its position id. Default is None.
+            kwargs (Dict): transformer layer inputs not relevant to cross attention.
+
+        Returns:
+            Tensor: output tensor with same shape as input
+                [batch_size x seq_length x embed_dim]
         """
+        # Skip cross attention when no secondary input
+        if encoder_input is None:
```
Comment (felipemello1) [inline torchtune/modules/transformer.py:181]: I find it a bit misleading. I dont think that someone should call a cross attention layer if they dont have anything to cross attend. IMO, it may lead to silent errors that are hard to debug, because it wont raise any errors. If the enconder_input is None, it should probably be resolved where the crossattention is being called.

File: torchtune/modules/transformer.py [hunk 2]
```diff
@@ -125,19 +251,28 @@ class TransformerDecoder(nn.Module):
 
     def __init__(
         self,
+        *,
         tok_embeddings: nn.Embedding,
-        layer: TransformerDecoderLayer,
-        num_layers: int,
+        layer: Union[nn.Module, List[nn.Module]],
```
Comment (felipemello1) [inline torchtune/modules/transformer.py:256]: do we need this flexibility, or can we simplify it by always creating the list internally?

File: torchtune/modules/transformer.py [hunk 3]
```diff
@@ -192,6 +329,10 @@ def forward(
                 before the softmax. A value of True in row i and column j means token i attends
                 to token j. A value of False means token i does not attend to token j. If no
                 mask is specified, a causal mask is used by default. Default is None.
+            encoder_input (Optional[Tensor]): Optional second input to cross attend with x. Shape
+                [batch_size x seq_length x embed_dim] (seq_length and embed_dim may very from x)
```
Comment (felipemello1) [inline torchtune/modules/transformer.py:333]: nit: very -> vary

File: torchtune/utils/_checkpointing/_checkpointer.py [hunk 4]
```diff
@@ -198,9 +198,17 @@ def load_checkpoint(self, weights_only: bool = True) -> Dict[str, Any]:
             Dict[str, Any]: state_dict from the input checkpoint
         """
         state_dict: Dict[str:Any] = {}
-        state_dict[utils.MODEL_KEY] = safe_torch_load(
+        model_state_dict = safe_torch_load(
             self._checkpoint_path, weights_only=weights_only
         )
+        if "layers.0.sa_norm.scale" in model_state_dict:
+            logger.info(
```
Comment (felipemello1) [inline torchtune/utils/_checkpointing/_checkpointer.py:205]: should it be logger.warning?

**Review by felipemello1 — COMMENTED** [review]
nice PR! I see its still a draft, but did a quick pass. Left some comments, mostly spelling / naming.

**Comment by pbontrager** [comment]
#1338 implements the fusion modules from here.
